source/common/singleton/threadsafe_singleton.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "absl/base/call_once.h"
+
+namespace Envoy {
+
+/**
+ * ThreadSafeSingleton allows easy global cross-thread access to a non-const object.
+ *
+ * This singleton class should be used for singletons which must be globally
+ * accessible but can not be marked as const. All functions in the singleton class
+ * *must* be threadsafe.
+ *
+ * Note that there is heavy resistence in Envoy to adding this type of singleton
+ * if data will persist with state changes across tests, as it becomes difficult
+ * to write clean unit tests if a state change in one test will persist into
+ * another test.  Be wary of using it.  A example of acceptable usage is OsSyscallsImpl,
+ * where the functions are not strictly speaking const, but affect the OS rather than the
+ * class itself.  An example of unacceptable usage upstream would be for
```

Comment:
Nit: double space after period.

Reply:
By design.  Is there a style guide line item for this?

Reply:
@dnoe also does this :) I think we agreed we would do single space, double space is some kind of crazy from the days of typewriters.